### PR TITLE
add cnn attention & simple loss logger

### LINKED ISSUE #4: Implement and integrate backbone Architecture: CNN + Multihead Attention
- could also change addition to CNN
- look at SleePyCo repo to find backbones

### PR DESCRIPTION
**Pull request by Kristen-Z**

### CODE REVIEW

**Review by RaphaelKreft — CHANGES_REQUESTED** [review]
The backbone needs a decoder for masked prediction. Please add one. You can have a look at #11 in cnn_backbone.py or ask @Ahwodafo

File: train_crl.py [hunk 1]
```diff
@@ -77,18 +81,24 @@ def train_one_epoch(self):
...
+        print(f"\n[INFO] Epoch {epoch}, Training Loss: {avg_train_loss:.4f}")
 
-            if self.train_iter % self.tp_cfg['val_period'] == 0:
-                print('')
-                val_loss = self.evaluate(mode='val')
-                self.early_stopping(None, val_loss, self.model)
-                self.model.train()
-                if self.early_stopping.early_stop:
-                    break
+        if self.train_iter % self.tp_cfg['val_period'] == 0:
```
Comment (RaphaelKreft) [inline train_crl.py]: by removing indentation this will evaluate at max 1 time at the end of an epoch but I don't think we want this? At least its not implemented like this in the repo.

File: train_crl.py [hunk 2]
```diff
@@ -104,17 +114,28 @@ def evaluate(self, mode):
...
+
+        avg_eval_loss = eval_loss / len(self.loader_dict[mode])
+        print(f"[INFO] {mode.capitalize()} Loss: {avg_eval_loss:.4f}")
 
-        return eval_loss
+        # Log validation loss to TensorBoard
+        if epoch is not None:
+            self.writer.add_scalar(f"Loss/{mode.capitalize()}", avg_eval_loss, epoch)
+
+        return avg_eval_loss
```
Comment (RaphaelKreft) [inline train_crl.py:136]: What are your thoughts regarding avg loss vs summed loss being returned here? @Kristen-Z

File: train_crl.py [hunk 3]
```diff
@@ -104,17 +114,28 @@ def evaluate(self, mode):
...
-            progress_bar(i, len(self.loader_dict[mode]), 'Lr: %.4e | Loss: %.3f' %(get_lr(self.optimizer), eval_loss / (i + 1)))
+            progress_bar(i, len(self.loader_dict[mode]), 'Lr: %.4e | Loss: %.3f' % (get_lr(self.optimizer), eval_loss / (i + 1)))
+
+        avg_eval_loss = eval_loss / len(self.loader_dict[mode])
+        print(f"[INFO] {mode.capitalize()} Loss: {avg_eval_loss:.4f}")
 
-        return eval_loss
+        # Log validation loss to TensorBoard
+        if epoch is not None:
+            self.writer.add_scalar(f"Loss/{mode.capitalize()}", avg_eval_loss, epoch)
```
Comment (RaphaelKreft) [inline train_crl.py]: If we let eval happen based on train iterations, log via train_iterations not epoch

**Review by RaphaelKreft — CHANGES_REQUESTED** [review]
also there are some changes to the logging necessary. You can look at #17 where I implemented and tested it for masked prediction training script

File: train_crl.py (hunk 2 above)
Comment (Kristen-Z) [reply train_crl.py:136]: I feel average loss is more standard, easy to compare, and batch size won’t skew it.

File: train_crl.py (hunk 1 above)
Comment (Kristen-Z) [reply train_crl.py]: You're right, my bad! It should be evaluated more often. Thanks for catching that!

File: train_crl.py (hunk 3 above)
Comment (Kristen-Z) [reply train_crl.py]: I have changed this and added more logging metrics in `train_ctrl.py` .

**Review by RaphaelKreft — APPROVED** [review]
Lgtm, will do integration. into existing code on another Branch
